turbinia/workers/analysis/ssh_analyzer.py

```diff
@@ -0,0 +1,397 @@
+      pyparsing.Literal(']:') + pyparsing.Literal('Disconnected') +
+      pyparsing.Literal('from') + pyparsing.Literal('user') + _USERNAME +
+      _SOURCE_IP + pyparsing.Literal('port') + _SOURCE_PORT)
+
+  MESSAGE_GRAMMAR = {
+      'accepted': _LOGIN_GRAMMAR,
+      'failed': _FAILED_GRAMMER,
+      'disconnected': _DISCONNECT_GRAMMAR
+  }
+
+  def read_logs(self, log_dir: str) -> pd.DataFrame:
+    """Read SSH authentication logs."""
+    ssh_records = []
+
+    for log_filename in os.listdir(log_dir):
+      if not log_filename.startswith(
+          'auth.log') and not log_filename.startswith('secure'):
+        continue
+
+      log_file = os.path.join(log_dir, log_filename)
+      log.debug(f'Processing authentication log {log_file}')
+
+      # Handle log archive
+      if log_filename.endswith('.gz'):
+        try:
+          with gzip.open(log_file, 'rt', encoding='ISO-8859–1') as fh:
+            log_data = fh.read()
+            records = self.read_log_data(log_data, log_filename=log_filename)
+            if records:
+              ssh_records += records
+        except gzip.BadGzipFile as e:
+          log.error(f'Error opening a bad gzip file {str(e)}')
+        finally:
+          continue
+
+      # Handle standard log file
+      try:
+        with open(log_file, 'r', encoding='ISO-8859–1') as fh:
+          log_data = fh.read()
+          records = self.read_log_data(log_data, log_filename=log_filename)
+          if records:
+            ssh_records += records
+      except FileNotFoundError:
+        log.error(f'{log_file} does not exist')
+      finally:
+        continue
+
+    if not ssh_records:
+      log.info(f'No SSH authenticaiton events in {log_dir}')
+      return pd.DataFrame()
+    log.info(
+        f'Total number of SSH authentication events {len(ssh_records)}'
+        f' in {log_dir}.')
+
+    ssh_data = []
+    for ssh_record in ssh_records:
+      ssh_data.append(ssh_record.__dict__)
+    df = pd.DataFrame(ssh_data)
+    return df
+
+  def parse_message_datetime(self, message_datetime, log_year):
+    """Parse and return datetime."""
+    # NOTE: returned datetime object contains naive datetime
+    # TODO(rmaskey): Better handle date time and timezone
+    if len(message_datetime) == 1:
+      return datetime.fromisoformat(message_datetime[0])
+    elif len(message_datetime) == 3:
+      datetime_string = (
+          f'{message_datetime[0]} {message_datetime[1]}'
+          f' {log_year} {message_datetime[2]}')
+      return datetime.strptime(datetime_string, '%b %d %Y %H:%M:%S')
+    else:
+      return datetime.fromtimestamp(0)
+
+  def read_log_data(
+      self, data, log_filename: str, log_year: int = None) -> List:
+    """ Parses SSH authentication log."""
+    # check valid year is provided
+    # If valid year isn't provided raise error
+    if not log_year:
+      current_date = datetime.now()
+      log_year = current_date.year
+      log.warning(
```

Comment:
I have added docstring with addional information.

The idea behind the `log_year` is because syslog does not record year in syslog events. So, `log_year` provides an option for users to set accurate year. However, this is likely not possible with Turbinia, and there is a code block that guesses log_year by comparing the last event with current date/time.

Reply:
It is used in function `read_log_data`.

Reply:
`log_year` value is passed to `parse_message_datetime` and it will be used when SSH messages are from Debian and Red Hat based systems.